### DISCUSSION TITLE
Kafkaconnect Deployment max surge and max unavailable

### DISCUSSION THREAD

**Discussion by ee07dazn** [strimzi/strimzi-kafka-operator#7883] [Q&A]
Hi 
Wish you a very Happy New Year!!

<img width="696" alt="image" src="https://user-images.githubusercontent.com/46483973/211307173-30bbd9f9-0bd3-435b-ab04-046ba99eef20.png">

As seen above, my default Kafkaconnect CRD creates a `deployment` with `RollingUpdateStrategy:  0 max unavailable, 1 max surge`. I would like to change this but on the [docs](https://strimzi.io/docs/operators/latest/configuring.html#type-DeploymentTemplate-reference)  I only see the option to specify the strategy.

Context - I am using [reloader](https://github.com/stakater/Reloader) to detect change in configmap and/or secrets to roll over the pods but this is not working as desired, for > 1 replica. I believe changing the above should help sort this issue for me.

**Comment by scholzj** [comment]
The docs are correct, you cannot configure it. I'm not sure I understand why would the `reloader` have problems with it. This is IMHO a regular configuration for a Deployment. In any case, the plan is to move away from using Deployments since they do not work well for Kafka Connect (see https://github.com/strimzi/proposals/blob/main/045-Stable-identities-for-Kafka-Connect-worker-nodes.md for more details).

**Comment by ee07dazn** [comment]
Thanks for sharing the proposal. 

I tested the reloader with the same settings on a different sample deployment and it works perfectly alright there. So something seems to be up on my strimzi Kafkaconnect. 

The strange behaviour I am noticing is 
when a configmap changes, reloader correctly correct triggers the deployment to rollover the pods 
However, 
1. the pods as well as the replicaset starts to rollover and then the pods come in running state.
2. but soon after, the deployment takes out the new replicaset (*ddf in image) and brings back the old replicaset (*b57 in image)
<img width="1192" alt="image" src="https://user-images.githubusercontent.com/46483973/211423509-3ff98a96-3579-4268-8327-acefbe619261.png">

Any idea what might be going on here ?

I have tried to look at the events and other things but i am not able to get a root cause of this issue.

**Reply by ee07dazn** [reply]
Any hint on where else to look for, will also be highly appreciated 🙏 as i have now ran out of ideas :(

**Reply by scholzj** [reply]
TBH, I never used this tool. So I'm just guessing. But from the screenshot and from [this](https://github.com/stakater/Reloader#reload-strategies) description it seems it modifies the Deployment resource. But after Strimzi noticed it, it will modify it back to correspond to the `KafkaConnect` CR. I guess that will get the pods rolled and the ConfigMaps / Secrets reloaded. But depending on the timing it might probably roll it twice and cause unnecessary disruption.

**Reply by ee07dazn** [reply]
Thanks!
That behaviour when coupled with more than 1 replica and RollingStrategy, actually leaves two pods in different state compared to the one that gets rolled over first. As by the time first one rolls over, Strimzi realises and reverts to the old rs and the old two pods remain unaffected. 

Using a Recreate strategy isn't great personally as it suddenly will delete all and recreate them again. 

Anyways, do you have any suggestion on how to roll over automatically when CMs or Secrets attached to the kafkaconnect CRD changes ? 

Manually, I could delete one pod at a time and roll them over, but I am sure there is a better approach even if reloader is not the optimal solution for KafkaConnect CR.

**Reply by scholzj** [reply]
I do not have any suggestions I'm afraid. 

Watching the secret for changes could be possibly implemented as a feature in Strimzi (it currently watches secrets used for authentication or TLS for example. But not the custom secrets or ConfigMaps users mounted into the Connect deployment for use in connectors). But I do not think everyone would want it, so it would need to be optional. 

Feel free to open an enhancement issue for it. Contributions are always welcomed, so if you want, you can contribute it ;-). But even if not, someone else might get to it if it is tracked.

**Reply by ee07dazn** [reply]
Sure thing. I will open an enhancement request. And will definitely try to take a stab at it. 


With the mounts, as per Kubernetes doc from 1.21 onwards, when a CM is mounted as volume (not when used as environment variable), and it changes, it get [auto updated](https://kubernetes.io/docs/concepts/configuration/configmap/#mounted-configmaps-are-updated-automatically) but i need to check for secrets.

Thanks for your suggestions. 🙏 

I will leave this open for a while, as i try to find a decent soln and post it here in case someone else needs it.

**Reply by scholzj** [reply]
> With the mounts, as per Kubernetes doc from 1.21 onwards, when a CM is mounted as volume (not when used as environment variable), and it changes, it get [auto updated](https://kubernetes.io/docs/concepts/configuration/configmap/#mounted-configmaps-are-updated-automatically) but i need to check for secrets.

AFAIK the Secrets update as well. but the problem is that Connect does not care about it. What the rolling of the pods does for you is not just update the ConfigMap or Secret content in the Pod. It also has to essentially restart the connectors and that is when they reload the data. So you would still need to trigger the connectors to reload the data somehow (e.g. restart them I guess).

**Reply by ee07dazn** [reply]
Ahhkk... I guess in some cases where the connector is not as much impacted, we should be good, right ?For ex- when changing the configMap for jmxprom etc. 

On that connector restart point, one of the old version docs had mentioned about annotations that should trigger a connector/task restart but that seems to have gone away from the more recent docs. 

I had also tried these annotations but they didn't work. Can you please confirm they are still valid ?

Kubectl annotate kafkaconnector <connector-name> strimzi.io/restart=true 

Kubectl annotate kafkaconnector <connector-name> Strimzi.io/restart-task=<task-number>
![image](https://user-images.githubusercontent.com/46483973/211691995-33bdd499-cc88-4851-8878-169f159268c6.jpeg)

**Reply by scholzj** [reply]
The operator IMHO should already today take care of any changes in the secrets it uses directly. The Prometheus metrics configuration is actually reloaded without restart. If for example some secret with password or certificate used for Kafka authentication changes, it will roll Connect on its own. But it will not do that for the secrets or config maps you mount for connectors.

The annotations should IMHO still work. If they don't, it would be a bug. But keep in mind that you need to have the Connector operator enabled and use the KafkaConnector resources. Otherwise, you need to restart them through the Connect REST API.

**Reply by ee07dazn** [reply]
i am definitely **not** using the REST API and only using the operator along with the resource. I will double check and revert back.

**Reply by ee07dazn** [reply]
> > With the mounts, as per Kubernetes doc from 1.21 onwards, when a CM is mounted as volume (not when used as environment variable), and it changes, it get [auto updated](https://kubernetes.io/docs/concepts/configuration/configmap/#mounted-configmaps-are-updated-automatically) but i need to check for secrets.
> 
> AFAIK the Secrets update as well. but the problem is that Connect does not care about it. What the rolling of the pods does for you is not just update the ConfigMap or Secret content in the Pod. It also has to essentially restart the connectors and that is when they reload the data. So you would still need to trigger the connectors to reload the data somehow (e.g. restart them I guess).

Hi 
So on that comment where you say you need to restart the connector possibly, this is specific when something that is used by the connector (say a secret) has been updated in the pod, which rolled over, right? otherwise the connector still uses the old config (say that secret) ?


I tried making change to 
1. a running connector
2. rolling over pods in general
and in both scenarios, the rebalancing happened and the connector worked fine eventually in about couple of mins, and whilst that was happening, my data was still coming through correctly. So all in all, the transition seemed smooth.

**Reply by scholzj** [reply]
When you roll the pod, the connectors (their tasks) start again. So yes, in that case they load any new changes automatically. But they would not do that if the pod doesn't restart.
